### PR TITLE
Initial add of logentries callback plugin

### PR DESCRIPTION
**Pull request by jcftang**
This callback plugin will generate json objects to be sent to the
logentries service for auditing/debugging purposes.

To use:

Add this to your ansible.cfg file in the defaults block

```
[defaults]
callback_plugins = ./callback_plugins
callback_stdout = logentries
callback_whitelist = logentries
```

Copy the callback plugin into the callback_plugings directory

Either set the environment variables

```
export LOGENTRIES_API=data.logentries.com
export LOGENTRIES_PORT=10000
export LOGENTRIES_ANSIBLE_TOKEN=dd21fc88-f00a-43ff-b977-e3a4233c53af
```

Or create a logentries.ini config file that sites next to the plugin with the following contents

```
[logentries]
api = data.logentries.com
port = 10000
tls_port = 20000
use_tls = no
token = dd21fc88-f00a-43ff-b977-e3a4233c53af
```

### CODE REVIEW

File: lib/ansible/modules/core [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit d659c79db7335a7a81594f7d977cb6bb67e9c5a3
+Subproject commit a4ac171a070551f1ed17a9c09b41d2481ad779af
```
Comment (bcoca) [inline lib/ansible/modules/core]: need to redo w/o the submodule updates

File: lib/ansible/plugins/callback/logentries.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
+"""
+(c) 2015, Logentries.com
+Author: Jimmy Tang <jimmy.tang@logentries.com>
+
```
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py]: needs gpl license

File: lib/ansible/plugins/callback/logentries.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    port = 10000
+    tls_port = 20000
+    use_tls = no
+    token = dd21fc88-f00a-43ff-b977-e3a4233c53af
+
+
+"""
+
+import os
+import threading
```
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py]: threading is pretty useless in this context, the callback gets called in it's own sub process

File: lib/ansible/plugins/callback/logentries.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                        continue
+                    break
+        except KeyboardInterrupt:
+            if self.verbose:
+                print("Logentries asynchronous socket client interrupted")
+
+        self.close_connection()
+
+
+try:
```
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py:164]: ~~ use the shared open_url function, it takes care of many issues with python's ssl ~~

File: lib/ansible/plugins/callback/logentries.py (hunk 3 above)
Comment (jcftang) [inline lib/ansible/plugins/callback/logentries.py]: Is the subprocess long lived (for the duration of the play). The code was lifted from an existing library that we have and we found it's a bit more reliable to have a queue to buffer the log lines before it's sent the logentries end points

File: lib/ansible/plugins/callback/logentries.py (hunk 2 above)
Comment (jcftang) [inline lib/ansible/plugins/callback/logentries.py]: Will update this once I check the my line manager

File: lib/ansible/plugins/callback/logentries.py (hunk 4 above)
Comment (jcftang) [inline lib/ansible/plugins/callback/logentries.py:164]: I had a look at the open_url function, it appears to be for http/https based connections only, the connection that this callback opens up is a plain tcp connection.

Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py:164]: ignore it then,  I stopped reading at import ssl, did not realize it is an encrypted tcp socket connection and assumed http/s

File: lib/ansible/plugins/callback/logentries.py (hunk 3 above)
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py]: this process does live for the lentgh of the play, but it is also not thread safe, we update variables that will end up being shared in the threads.

Comment (jcftang) [inline lib/ansible/plugins/callback/logentries.py]: Ok, I will work on removing the threading then.

**Comment by jcftang** [comment]
I've removed the threading and queues, added a license as required.

I've also cleaned up a bit with wrapping up the emit method.

File: lib/ansible/plugins/callback/logentries.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+import os
+import socket
+import random
+import time
+import codecs
+import ConfigParser
+import uuid
+try:
+    import certifi
```
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py:66]: import errors should not be fatal, look at hipchat plugin and how it handles this, the plugin will not work but will not stop play while it issues warnings about it.

File: lib/ansible/plugins/callback/logentries.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+            self._conn = sock
+
+    SocketAppender = TLSSocketAppender
+
+
+class CallbackModule(CallbackBase):
+    CALLBACK_VERSION = 2.0
+    CALLBACK_TYPE = 'notification'
+    CALLBACK_NAME = 'logentries'
+
```
Comment (bcoca) [inline lib/ansible/plugins/callback/logentries.py:196]: add 
 CALLBACK_NEEDS_WHITELIST = True

this is recent addition, it keeps plugins shipped with ansible from being executed automatically unless whitelisted through config.

File: lib/ansible/plugins/callback/logentries.py (hunk 5 above)
Comment (jcftang) [inline lib/ansible/plugins/callback/logentries.py:66]: Sounds reasonable, thats changed in the following commits

**Pull request merged**